graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,332 @@
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        self.fit_predict(X)
+        return self
+
+    def fit_predict(self, X, level=None):
+        """
+        Fits clustering models to the data as well as resulting clusters
+        and using fitted models to predict a hierarchy of labels
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+        level: int, optional (default=None)
+
+        Returns
+        -------
+        labels : array_label, shape (n_samples, n_levels)
+            if no level specified; otherwise, shape (n_samples,)
+        """
+        X = check_array(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+        if level is not None:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+            elif level < 1:
+                raise ValueError("level must be positive")
+
+        if self.max_components > X.shape[0]:
+            msg = "max_components must be >= n_samples, but max_components = "
+            msg += "{}, n_samples = {}".format(self.max_components, X.shape[0])
+            raise ValueError(msg)
+
+        labels = self._fit(X)
+        labels = self._relabel(labels)
+        if level is not None:
+            if level <= labels.shape[1]:
+                labels = labels[:, level - 1]
+            else:
+                msg = "input exceeds max level = {}".format(labels.shape[1])
+                raise ValueError(msg)
+
+        return labels
+
+    def _cluster_and_decide(self, X):
+        if self.is_root:
+            min_components = self.min_components
+        else:
+            min_components = 1
+
+        if self.cluster_method == "gmm":
+            cluster = AutoGMMCluster(
+                min_components=min_components,
+                max_components=self.max_components,
+                **self.cluster_kws
+            )
+            cluster.fit(X)
+            model = cluster.model_
+            criter = cluster.criter_
+            k = cluster.n_components_
+            pred = cluster.predict(X)
+
+            if self.delta_criter:
+                single_cluster = AutoGMMCluster(
+                    min_components=1, max_components=1, **self.cluster_kws
+                )
+                single_cluster.fit(X)
+                criter_single_cluster = single_cluster.criter_
+
+                if k > 1:
+                    # check whether the difference between the criterion
+                    # of "split" and "not split" is greater than
+                    # the threshold, delta_criter
+                    if criter_single_cluster - criter < self.delta_criter:
+                        k = 1
+                        pred = np.zeros((len(X), 1), dtype=int)
+
+        elif self.cluster_method == "kmeans":
```

Comment:
@bdpedigo Something I forgot to ask while ago: the min_comp is always 2 for `KMeansCluster`, so that clustering only stops if too few samples are left or current level is too deep. Do you think that's a problem because of inconsistency w/ gmm?

Reply:
@tliu68 can you explain what you mean by min_comp is always 2? I know that is the default. But KMeans should still run for k=1 (I think, based on a quick look at the code, i do know it's a bit silly). are you talking about the fact that silhouette_score doesnt make sense for k=1?

Reply:
Yes, I was talking about since silhouette_score is not defined for k=1, kmeans won't decide not to split b/c the current model has the best silhouette_score compared to models of more clusters.